pkg/typesystem/connected_types.go

```diff
@@ -0,0 +1,123 @@
+package typesystem
+
+import (
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	"github.com/openfga/openfga/pkg/tuple"
+)
+
+// map[_objectType_]map[_relation_]map[_subjectType_] terminalRelation.
+type TypesystemConnectedTypes map[string]map[string]map[string][]string
+
+func (f TypesystemConnectedTypes) assign(objectType string, relation string, subjectType string, terminalRelation string) {
+	if f[objectType] == nil {
+		f[objectType] = make(map[string]map[string][]string)
+	}
+
+	if f[objectType][relation] == nil {
+		f[objectType][relation] = make(map[string][]string)
+	}
+
+	f[objectType][relation][subjectType] = append(f[objectType][relation][subjectType], terminalRelation)
+}
+
+// AssignTerminalTypes will populate the `connectedTypes` property on the typesystem to indicate for a given object type
+// what the terminal user types and relations.
+func (t *TypeSystem) AssignTerminalTypes(typeName, relationName string) {
+	terminalTypesAndRelations := t.getTerminalUserTypeAndRelationsForConnectedTypes(typeName, relationName, 0)
+	for _, terminalTypesAndRelation := range terminalTypesAndRelations {
+		for _, terminalType := range terminalTypesAndRelation.terminalTypes {
+			t.connectedTypes.assign(typeName, relationName, terminalType, terminalTypesAndRelation.terminalRelation)
+		}
+	}
+}
+
+type terminalTypesAndRelation struct {
+	terminalTypes    []string
+	terminalRelation string
+}
+
+func (t *TypeSystem) getTerminalUserTypeAndRelationsForConnectedTypes(
```

Comment:
This function is the core of the PR. During typesystem create, recursively traverses it to determine a mapping of all the terminal types and terminal relations. As more scenarios are supported in the future, this function will need to be updated.